### ISSUE TITLE
Spurious error for code that fails truthy-bool even when not enabled

### ISSUE THREAD

**Issue by huonw** [python/mypy#18026]
**Bug Report**

It seems that code failing to pass the `truthy-bool` check can have spurious errors that occur, even when that optional error code isn't enabled.

**To Reproduce**

https://mypy-play.net/?mypy=latest&python=3.12&gist=53000744586f03d23bb53f807d5a0ea0

```python
class Truthy: 
    pass
    # def __bool__(self) -> bool: return True

def foo(truthies: list[Truthy]) -> None:
    sum(1 if x else 0 for x in truthies)

```

**Expected Behavior**

This code should typecheck fine: `sum` on `int`s is sensible!

Uncommenting the `__bool__` definition (so that the `truthy-bool` error no longer applies, if it was enabled) passes type checking.

**Actual Behavior**

```
main.py:4: error: Generator has incompatible item type "int"; expected "bool"  [misc]
```

(Note that the `truthy-bool` error doesn't occur, as expected, because it isn't enabled.)

**Your Environment**

- Mypy version used: 1.12.1
- Mypy command-line flags: N/A see playground
- Mypy configuration options from `mypy.ini` (and other config files): N/A see playground
- Python version used: 3.12

**Comment by huonw** [comment]
Ah okay, I did some hacking around (narrowed the problem down to `Errors._filter_error` and the `ErrorWatcher`s, and so added unique stable index to each `ErrorWatcher`, and then binary-searched returning early from `on_error` for only specific instances (returning early stopped the bad behaviour), to find which exact one was of interest, and then printed the stack trace of its construction to find where).

This showed that the example in the description is caused by the hidden error interfering with return-type overload resolution:

https://github.com/python/mypy/blob/94c49a8215c9c8d3a4ec146696213dd2b9ebfb42/mypy/checkexpr.py#L2861-L2873

In particular, it seems like an overload is rejected if there's any error, even if it's one like this that's not related to type-selection (per se), and even if the error is not emitted.

This happens in the `Errors._filter_error` call, that calls `ErrorWatcher.on_error`. The code above is creating the `ErrorWatcher` in `self.msg.filter_errors()`.

https://github.com/python/mypy/blob/94c49a8215c9c8d3a4ec146696213dd2b9ebfb42/mypy/errors.py#L494-L499


Without knowing much about mypy, it seems like some errors should influence overload resolution, and other errors should not, and they should be distinguished, somehow.

This indicates to me that _any_ error seems to be enough to cause problems, e.g. this code: https://mypy-play.net/?mypy=latest&python=3.12&gist=602669a7523d13f2bfc0191f48cd54e1

```python
def foo(x: int) -> list[int]:
    return []
    
def bar() -> int:
    return sum(foo("x"))
```

```
main.py:5: error: Argument 1 to "sum" has incompatible type "list[int]"; expected "Iterable[bool]"  [arg-type]
main.py:5: error: Argument 1 to "foo" has incompatible type "str"; expected "int"  [arg-type]
```

The second error is correct (indeed `"x"` isn't valid to pass to `foo`), but the first is spurious: `sum` should work fine with `list[int]`. (This one isn't so bad, given it's an spurious error happening along with another real error.)

**Comment by huonw** [comment]
Ah, this looks to be the same cause as https://github.com/python/mypy/issues/17082, although the example in this issue is a more egregious instance, IMO.
